File: 백준/Silver/1495. 기타리스트/기타리스트.py
import sys

input = sys.stdin.readline
n, s, m = map(int, input().split())
v = list(map(int, input().split()))
d = [[0] * (m + 1) for _ in range(n + 1)]
d[0][s] = 1

for i in range(1, n + 1):
    for j in range(m + 1):
        if d[i - 1][j] == 1:
            if j - v[i - 1] >= 0:
                d[i][j - v[i - 1]] = 1
            if j + v[i - 1] <= m:
                d[i][j + v[i - 1]] = 1

result = -1
for i in range(m, -1, -1):
    if d[n][i] == 1:
        result = i
        break

print(result)

File: 백준/Silver/1495. 기타리스트/기타리스트.c
#include <stdio.h>

int main() {
    int n, s, m;
    scanf("%d %d %d", &n, &s, &m);

    int v[101]; // 곡 수 n 최대 100
    for (int i = 0; i < n; i++) {
        scanf("%d", &v[i]);
    }

    int d[101][1001] = {0}; // n은 최대 100, m은 최대 1000
    d[0][s] = 1; // 시작 볼륨

    for (int i = 1; i <= n; i++) {
        for (int j = 0; j <= m; j++) {
            if (d[i - 1][j] == 1) {
                if (j - v[i - 1] >= 0) {
                    d[i][j - v[i - 1]] = 1;
                }
                if (j + v[i - 1] <= m) {
                    d[i][j + v[i - 1]] = 1;
                }
            }
        }
    }

    int result = -1;
    for (int i = m; i >= 0; i--) {
        if (d[n][i] == 1) {
            result = i;
            break;
        }
    }

    printf("%d\n", result);
    return 0;
}